[JSC] Use GPRInfo::nonPreservedNonArgumentGPR1 for BBQJIT boxedCallee GPR

Constellation opened the pull request:
#### 821c07448c2f82f0fd166d1090b6f01d8a9acd08
<pre>
[JSC] Use GPRInfo::nonPreservedNonArgumentGPR1 for BBQJIT boxedCallee GPR
<a href="https://bugs.webkit.org/show_bug.cgi?id=300932">https://bugs.webkit.org/show_bug.cgi?id=300932</a>
<a href="https://rdar.apple.com/162807925">rdar://162807925</a>

Reviewed by Yijia Huang.

This patch pins BBQJIT boxedCallee GPR as GPRInfo::nonPreservedNonArgumentGPR1.
To introduce polymorphic call profiling, we will introduce a bit more
complex code to count polymorphic cases. This will require a shared
thunk which sometimes calls C++ function to allocate polymorphic call
profile data structure. Thus we need to ensure that boxedCallee are
always placed in a particular register after we set up all wasm arguments.
This means it needs to be in non-preserved non-argument register. Also
we avoid using GPRInfo::nonPreservedNonArgumentGPR0 since it is already
used for wasmScratchGPR. Thus we pick GPRInfo::nonPreservedNonArgumentGPR1.

With this, we also move the current profiling code from &quot;before setting
up the arguments&quot; to &quot;after setting up the arguments&quot; since this ensures
that we can call a function if we save and restore arguments registers.
This should rarely happen (only when we allocate polymorphic call
profiling data structure allocation), thus it is costly but fine.

* Source/JavaScriptCore/assembler/MacroAssembler.h:
(JSC::MacroAssembler::orPtr):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitIndirectCall):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitIndirectTailCall):
(JSC::Wasm::BBQJITImpl::BBQJIT::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addCallRef):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addCallRef):
* Source/JavaScriptCore/wasm/WasmCallProfile.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
(JSC::WebAssemblyFunctionBase::offsetOfImportableFunction):

Canonical link: <a href="https://commits.webkit.org/301819@main">https://commits.webkit.org/301819@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a803110022513a44e9925af3f3f8a5a5336b04a0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127184 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46820 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37960 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134248 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/78739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2a5356df-bea3-4c75-960f-d958d975b00d) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47436 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55346 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/96790 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/78739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1e285709-3a89-4896-bc12-bf70e3e33833) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130132 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/37980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/113917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77295 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/65b97299-0651-43df-a9e6-5ed693531f5b) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/36873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/32059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77628 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/119221 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/107844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32401 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136731 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125647 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/53839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/41486 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105310 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/54345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110270 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104998 "Passed tests") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/28984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/51406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19892 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/53780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59867 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158684 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/53012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39688 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/56345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/54773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian reviewed (commented):
typo in commit message: "that we can all a function".  I think you meant to say "call" instead of "all".

hyjorc1 reviewed (approved):
r=me

MenloDorian commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -4489,57 +4489,65 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addCall(unsigned callProfileIndex, Func
+
     if (m_profile->isMegamorphic(callProfileIndex)) {
+        m_jit.loadPtr(CCallHelpers::Address(importableFunction, WasmToWasmImportableFunction::offsetOfBoxedCallee()), wasmScratchGPR);
+        m_jit.storeWasmCalleeToCalleeCallFrame(wasmScratchGPR);
```
These 2 lines are identical between the `if` and `else` case.  Can you hoist them out?

Constellation replied:
Done

Constellation committed:
[JSC] Use GPRInfo::nonPreservedNonArgumentGPR1 for BBQJIT boxedCallee GPR
https://bugs.webkit.org/show_bug.cgi?id=300932
rdar://162807925

Reviewed by Yijia Huang.

This patch pins BBQJIT boxedCallee GPR as GPRInfo::nonPreservedNonArgumentGPR1.
To introduce polymorphic call profiling, we will introduce a bit more
complex code to count polymorphic cases. This will require a shared
thunk which sometimes calls C++ function to allocate polymorphic call
profile data structure. Thus we need to ensure that boxedCallee are
always placed in a particular register after we set up all wasm arguments.
This means it needs to be in non-preserved non-argument register. Also
we avoid using GPRInfo::nonPreservedNonArgumentGPR0 since it is already
used for wasmScratchGPR. Thus we pick GPRInfo::nonPreservedNonArgumentGPR1.

With this, we also move the current profiling code from "before setting
up the arguments" to "after setting up the arguments" since this ensures
that we can call a function if we save and restore arguments registers.
This should rarely happen (only when we allocate polymorphic call
profiling data structure allocation), thus it is costly but fine.

* Source/JavaScriptCore/assembler/MacroAssembler.h:
(JSC::MacroAssembler::orPtr):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitIndirectCall):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitIndirectTailCall):
(JSC::Wasm::BBQJITImpl::BBQJIT::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addCallRef):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addCallRef):
* Source/JavaScriptCore/wasm/WasmCallProfile.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
(JSC::WebAssemblyFunctionBase::offsetOfImportableFunction):

Canonical link: https://commits.webkit.org/301819@main
